server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -540,13 +559,81 @@ private static SnapshotsInRepo sortAndFilterSnapshots(
             final Predicate<SnapshotInfo> fromSortValuePredicate = buildFromSortValuePredicate(sortBy, fromSortValue, order, null, null);
             if (predicate == null) {
                 predicate = fromSortValuePredicate;
-            } else {
+            } else if (fromSortValuePredicate != null) {
                 predicate = fromSortValuePredicate.and(predicate);
             }
         }
         return predicate;
     }
 
+    // Builds a predicate that can be applied to a combination of snapshot id and repository data to filter out snapshots that are not
+    // required to answer the request before loading their SnapshotInfo from the repository
+    // TODO: extend this method to cover the pagination after value as well where possible
+    @Nullable
+    private static BiPredicate<SnapshotId, RepositoryData> buildSnapshotPreflightPredicate(
```

Comment:
Can we instead have a single method that builds both the early and the late filter? I think that would make it easier to instinctly verify that the return nulls in `buildFromSortValuePredicate` are correct.

Reply:
I made it build both predicates in the same code path now to make this easier to follow :)